src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                    && !usePSK
+#endif
+                    )
+            {
+                TLSX_Remove(&parsedExts, TLSX_KEY_SHARE, ssl->heap);
+            }
+
+            {
+                WOLFSSL* nonConstSSL = (WOLFSSL*)ssl;
+                TLSX* sslExts = nonConstSSL->extensions;
+
+                /* This is required to correctly generate the hash */
+                ret = SetCipherSpecs_ex(WOLFSSL_SERVER_END, cs.cipherSuite0,
+                                     cs.cipherSuite, &nonConstSSL->specs, NULL);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                nonConstSSL->options.tls = 1;
+                nonConstSSL->options.tls1_1 = 1;
+                nonConstSSL->options.tls1_3 = 1;
+
+                XMEMCPY(nonConstSSL->session->sessionID, ch->sessionId.elements,
+                        ch->sessionId.size);
+                nonConstSSL->session->sessionIDSz = ch->sessionId.size;
+                nonConstSSL->options.cipherSuite0 = cs.cipherSuite0;
+                nonConstSSL->options.cipherSuite = cs.cipherSuite;
+                nonConstSSL->extensions = parsedExts;
+
+
+                ret = SendTls13ServerHello(nonConstSSL, hello_retry_request);
+
+                InitCipherSpecs(&nonConstSSL->specs);
+
+                nonConstSSL->session->sessionIDSz = 0;
+                nonConstSSL->options.cipherSuite0 = 0;
+                nonConstSSL->options.cipherSuite = 0;
+                nonConstSSL->extensions = sslExts;
+
+                nonConstSSL->options.tls = 0;
+                nonConstSSL->options.tls1_1 = 0;
+                nonConstSSL->options.tls1_3 = 0;
+            }
+dtls13_cleanup:
+            TLSX_FreeAll(parsedExts, ssl->heap);
+        }
+        else
+            ret = SendAlert((WOLFSSL*)ssl, alert_fatal, illegal_parameter);
+#else
+#error "WOLFSSL_SEND_HRR_COOKIE has to be defined to use DTLS 1.3 server"
+#endif
+    }
+    else
+#endif
+    {
+        if (!ch->dtls12cookieSet) {
+            ret = CreateDtls12Cookie(ssl, ch, ch->dtls12cookie);
+            if (ret != 0)
+                return ret;
+            ch->dtls12cookieSet = 1;
+        }
+        ret = SendHelloVerifyRequest((WOLFSSL*)ssl, ch->dtls12cookie,
+                DTLS_COOKIE_SZ);
+    }
+    return ret;
+}
+
+int DoClientHelloStateless(WOLFSSL* ssl, const byte* input,
+                           word32* inOutIdx, word32 helloSz)
 {
-    byte cookie[DTLS_COOKIE_SZ];
     int ret;
     WolfSSL_CH ch;
+    byte isTls13 = 0;
+    PskInfo pskInfo;
 
-    *process = 1;
+    XMEMSET(&pskInfo, 0, sizeof(pskInfo));
+    XMEMSET(&ch, 0, sizeof(ch));
+
+    ssl->options.dtlsStateful = 0;
     ret = ParseClientHello(input + *inOutIdx, helloSz, &ch);
     if (ret != 0)
         return ret;
 
 #ifdef WOLFSSL_DTLS_NO_HVR_ON_RESUME
-    {
-        byte isValid = 0;
-        ret = TlsResumptionIsValid(ssl, &ch, &isValid);
+    ret = TlsResumptionIsValid(ssl, &ch, &pskInfo);
```

Comment:
I don't think it depends on the TLS 1.3 check. Is there anything specific you have in mind?

Reply:
IIRC now we always do a cookie check in dtls 1.3, we do not have the `DTLS_NO_HVR_ON_RESUME` feature in dtls 1.3. 
If, for other reasons, we always need to check about the session in dtls 1.3 this shouldn't be guarded by `WOLFSSL_DTLS_NO_HVR_ON_RESUME`.
If the logic from 1.3 and 1.2 starts diverging too much it may be worth considering refactoring the function in two to make it simpler.

Reply:
What do you think of using `WOLFSSL_DTLS_NO_HVR_ON_RESUME` in 1.3 to mean no HRR on resume?

Reply:
I think extending `WOLFSSL_DTLS_NO_HVR_ON_RESUME` to 1.3 to mean the same thing as 1.2 is easier to understand. This is disabled by default so the user must declare it explicitly anyway.